### PR TITLE
Subgraph manifest template

### PR DESCRIPTION
**Pull request by danielattilasimon**
Quick and lightweight implementation of templating for the subgraph manifest in pure JS, without using a library. Can work with syntax hightlight 😉 

(I used "YAML for JavaScript" in VS Code).

image

Future improvements:
- add a parameter for choosing the network (mainnet/dev/kovan...)
- add startingBlock to deployment manifests to eliminate hardcoding
- split out the (very minimal) logic that renders the template into a separate file

### CODE REVIEW

File: packages/subgraph/src/entities/Token.ts [hunk 1]
```diff
@@ -1,10 +1,10 @@
 import { Address } from "@graphprotocol/graph-ts";
 
 import { Token } from "../../generated/schema";
-import { ERC20 } from "../../generated/templates/Token/ERC20"
+import { ERC20 } from "../../generated/LUSDToken/ERC20";
```
Comment (bingen) [inline packages/subgraph/src/entities/Token.ts:4]: Why this change? I used the one in template so this generic entity wouldn’t rely on an specific token code.

File: packages/subgraph/src/entities/TokenAllowance.ts [hunk 2]
```diff
@@ -6,8 +6,12 @@ import { TokenAllowance } from "../../generated/schema";
...
 import { getUser } from "./User";
 
-export function getTokenAllowance(_token: Address, _owner: Address, _spender: Address): TokenAllowance {
-  let id = _token.toHexString() + _owner.toHexString() + _spender.toHexString();
+export function getTokenAllowance(
+  _token: Address,
+  _owner: Address,
+  _spender: Address
+): TokenAllowance {
+  let id = _token.toHexString() + "-" + _owner.toHexString() + "-" + _spender.toHexString();
```
Comment (bingen) [inline packages/subgraph/src/entities/TokenAllowance.ts:14]: Should we set up a linter so we avoid these kind of changes?

File: packages/subgraph/src/entities/TokenBalance.ts [hunk 3]
```diff
@@ -27,11 +27,17 @@ export function getTokenBalance(_token: Address, _owner: Address): TokenBalance
...
+  _from: Address,
+  _to: Address,
+  _value: BigInt
+): void {
   let tokenAddress = _event.address;
   let token = getToken(tokenAddress);
 
-  if (_from.toHexString() == ZERO_ADDRESS) { // mint
+  if (_from.toHexString() == ZERO_ADDRESS) {
+    // mint
```
Comment (bingen) [inline packages/subgraph/src/entities/TokenBalance.ts:40]: Was this automated by the editor or it’s that you like it better this way?

File: packages/subgraph/subgraph.yaml.js [hunk 4]
```diff
@@ -1,18 +1,32 @@
+const fs = require("fs");
+
+const { addresses } = require("@liquity/lib-ethers/deployments/mainnet.json");
+// const { addresses } = require("@liquity/lib-ethers/deployments/dev.json");
```
Comment (bingen) [inline packages/subgraph/subgraph.yaml.js:4]: It would be nice to be able to pass this as a param, or have something like `prepare-local` and `prepare`, or whatever, so we don’t have to manually uncomment this.

File: packages/subgraph/subgraph.yaml.js [hunk 5]
```diff
@@ -1,18 +1,32 @@
...
+
+// https://etherscan.io/tx/0x0b612c6ffcef059a1cb9ceda83dee44a3d74e35d5018f1b8b486f3186cd7850e
+const startBlock = 12178551;
+// const startBlock = 0;
+
+const yaml = (strings, ...keys) =>
```
Comment (bingen) [inline packages/subgraph/subgraph.yaml.js:10]: Why do we need this and the tagged templates? Wouldn’t just work if we simply define:
```
const manifest = `
specVersion: 0.0.2
(...)
      address: "${addresses.troveManager}"
      startBlock: ${startBlock}
      (...)
`
```
The variables would already be replaced, right?

File: packages/subgraph/src/mappings/TroveManager.ts [hunk 6]
```diff
@@ -4,20 +4,8 @@ import {
   TroveLiquidated,
   Liquidation,
   Redemption,
-  BorrowerOperationsAddressChanged,
-  StabilityPoolAddressChanged,
-  CollSurplusPoolAddressChanged,
-  PriceFeedAddressChanged,
-  LQTYStakingAddressChanged,
-  LUSDTokenAddressChanged
```
Comment (bingen) [inline packages/subgraph/src/mappings/TroveManager.ts:12]: What’s the benefit of dropping this and adding the addresses to `subgraph.yaml`? Isn’t it more error prone? For instance if in some environment we have 2 deployments (not on mainnet so far, but it could happen locally or on a testnet), and for some reason one of the addresses is not changed, then the subgraph would be all messed up.

**Review by bingen — APPROVED** [review]
Thanks for tackling this! It was on my TODO list :wink:

File: packages/subgraph/subgraph.yaml.js (hunk 5 above)
Comment (edmulraney) [reply packages/subgraph/subgraph.yaml.js:10]: I think the whole thing would be simpler as just a tagged template literal

**Review by edmulraney — APPROVED** [review]

File: packages/subgraph/src/entities/Token.ts (hunk 1 above)
Comment (danielattilasimon) [reply packages/subgraph/src/entities/Token.ts:4]: LUSDToken and LQTYToken are concrete data sources in the manifest now. There's no longer a template.

File: packages/subgraph/src/entities/TokenBalance.ts (hunk 3 above)
Comment (danielattilasimon) [reply packages/subgraph/src/entities/TokenBalance.ts:40]: This is automatically done by prettier. We have a config file in the root dir:
https://github.com/liquity/dev/blob/main/.prettierrc.json

File: packages/subgraph/subgraph.yaml.js (hunk 5 above)
Comment (danielattilasimon) [reply packages/subgraph/subgraph.yaml.js:10]: Yeah, I didn't want to lose syntax highlighting.

Comment (bingen) [reply packages/subgraph/subgraph.yaml.js:10]: Oh, I see! Cool, thanks for the clarification!

File: packages/subgraph/src/entities/TokenBalance.ts (hunk 3 above)
Comment (bingen) [reply packages/subgraph/src/entities/TokenBalance.ts:40]: Oh, nice! I’ll try to remember to run it always before pushing (maybe I’ll try to add a git hook next time, as I’m kind of forgetful, haha).
Thanks!

File: packages/subgraph/src/mappings/TroveManager.ts (hunk 6 above)
Comment (danielattilasimon) [reply packages/subgraph/src/mappings/TroveManager.ts:12]: I only used these events because we didn't have a template for the manifest, so I tried to hardcode as few addresses into the manifest as I could. That's no longer an issue, so it seems pointless to have all these boilerplate handlers in the code. It's impossible to get these events more than once anyway, because ownership is revoked upon setting them.

I'm not sure I get why having the addresses in the manifest is more error prone.

File: packages/subgraph/src/entities/Token.ts (hunk 1 above)
Comment (bingen) [reply packages/subgraph/src/entities/Token.ts:4]: Right, I forgot to get back to this comment when I saw the changes below.

File: packages/subgraph/src/mappings/TroveManager.ts (hunk 6 above)
Comment (bingen) [reply packages/subgraph/src/mappings/TroveManager.ts:12]: > I'm not sure I get why having the addresses in the manifest is more error prone.

Precisely because of this:
> It's impossible to get these events more than once anyway, because ownership is revoked upon setting them.

:wink: 
With the events you make sure all addresses are consistent. The way it is now, you rely on the addresses on the manifest, and although we have a script to replace them, the source may be wrong (for instance the `dev.json` file). But anyway, no big deal, I’m aware those json files are automatically generated too, so I guess it’s fine as it is.

File: packages/subgraph/src/entities/TokenAllowance.ts (hunk 2 above)
Comment (bingen) [reply packages/subgraph/src/entities/TokenAllowance.ts:14]: I meant the line above, but this is resolved with the comment about prettier config file below too, thanks!

File: packages/subgraph/src/mappings/TroveManager.ts (hunk 6 above)
Comment (danielattilasimon) [reply packages/subgraph/src/mappings/TroveManager.ts:12]: I also realized that some constuctors emit events that the subgraph may be interested in (if not now, then in the future). A prominent example is the LQTYToken constructor, which mints the entire LQTY supply to a few addresses.

If we were to instantiate the data sources for these contracts from templates upon receiving some event from a TX on another contract, we might miss some crucial events.

Comment (bingen) [reply packages/subgraph/src/mappings/TroveManager.ts:12]: Yes, that’s a good point, actually that’s why I had to move out LQTY token out from the generic template to an entity in my original tokens PR.

File: packages/subgraph/subgraph.yaml.js (hunk 4 above)
Comment (danielattilasimon) [reply packages/subgraph/subgraph.yaml.js]: Done 👍

Comment (bingen) [reply packages/subgraph/subgraph.yaml.js]: Awesome, thanks!
